src/Microsoft.AspNetCore.Blazor.Razor.Extensions/ComponentDocumentClassifierPass.cs

```diff
@@ -70,6 +76,38 @@ protected override bool IsMatch(RazorCodeDocument codeDocument, DocumentIntermed
             method.Children.Insert(0, callBase);
         }
 
+        // In general documents will have a relative path (relative to the project root).
+        // We can only really compute a nice class/namespace when we know a relative path.
+        //
+        // However all kinds of thing are possible in tools. We shouldn't barf here if the document isn't 
+        // set up correctly.
+        private bool TryComputeNamespaceAndClass(string relativePath, out string @namespace, out string @class)
+        {
+            if (relativePath == null)
+            {
+                @namespace = null;
+                @class = null;
+                return false;
+            }
+
+            var builder = new StringBuilder();
+            builder.Append(BaseNamespace); // Don't sanitize, we expect it to contain dots.
+
+            var segments = relativePath.Split(PathSeparators);
+
+            // Skip the last segment because it's the FileName.
+            for (var i = 0; i < segments.Length - 1; i++)
+            {
+                builder.Append('.');
+                builder.Append(CSharpIdentifier.SanitizeClassName(segments[i]));
+            }
+
+            @namespace = builder.ToString();
+            @class = Path.GetFileNameWithoutExtension(relativePath);
+
+            return true;
```

Comment:
This is very similar to your existing code, and MVC's implementation of `@namespace`